File: algorithms/sorting/pigeonhole-sort/python/pigeonhole_sort.py
def pigeonhole_sort(arr: list[int]) -> list[int]:
    """
    Pigeonhole Sort implementation.
    Efficient for sorting lists of integers where the number of elements is roughly the same as the number of possible key values.
    """
    if len(arr) == 0:
        return []

    min_val = min(arr)
    max_val = max(arr)
    size = max_val - min_val + 1

    holes: list[list[int]] = [[] for _ in range(size)]

    for x in arr:
        holes[x - min_val].append(x)

    result: list[int] = []
    for hole in holes:
        result.extend(hole)

    return result

File: algorithms/sorting/pigeonhole-sort/rust/pigeonhole_sort.rs
/**
 * Pigeonhole Sort implementation.
 * Efficient for sorting lists of integers where the number of elements is roughly the same as the number of possible key values.
 */
pub fn pigeonhole_sort(arr: &[i32]) -> Vec<i32> {
    if arr.is_empty() {
        return Vec::new();
    }

    let min_val = *arr.iter().min().unwrap();
    let max_val = *arr.iter().max().unwrap();
    let range = (max_val - min_val + 1) as usize;

    let mut holes: Vec<Vec<i32>> = vec![Vec::new(); range];

    for &x in arr {
        holes[(x - min_val) as usize].push(x);
    }

    let mut result = Vec::with_capacity(arr.len());
    for hole in holes {
        result.extend(hole);
    }

    result
}
